oss-internship-2020/curl/tests/test_utils.cc

```diff
@@ -0,0 +1,287 @@
+  if (!response.ok()) {
+    return response.status();
+  }
+
+  return response.value();
+}
+
+// Read the socket until str is completely read
+std::string ReadUntil(const int socket, const std::string& str,
+                      const size_t max_request_size) {
+  char buf[max_request_size] = {};
+  size_t read_bytes = 0;
+
+  // Read one char at a time until str is suffix of buf
+  do {
+    if (read_bytes >= max_request_size ||
+        read(socket, buf + read_bytes, 1) == -1) {
+      return "";
+    }
+    ++read_bytes;
+  } while (std::string{buf + std::max(size_t{0}, read_bytes - str.size())} !=
+           str);
+
+  buf[read_bytes] = '\0';
+  return std::string{buf};
+}
+
+// Parse HTTP headers to return the Content-Length
+size_t GetContentLength(const std::string& headers) {
+  // Find the Content-Length header
+  const char* length_header_start = strstr(headers.c_str(), "Content-Length: ");
+
+  // There is no Content-Length field
+  if (!length_header_start) {
+    return 0;
+  }
+
+  // Find Content-Length string
+  const char* length_start = length_header_start + strlen("Content-Length: ");
+  size_t length_bytes = strstr(length_start, "\r\n") - length_start;
+  if (length_bytes >= 64) {
+    return 0;
+  }
+
+  // Convert Content-Length string value to int
+  char content_length_string[64];
+  strncpy(content_length_string, length_start, length_bytes);
+
+  return atoi(content_length_string);
+}
+
+// Read exactly content_bytes from the socket
+std::string ReadExact(int socket, size_t content_bytes) {
+  char buf[content_bytes + 1] = {};
+  size_t read_bytes = 0;
+
+  // Read until content_bytes chars are read
+  do {
+    int num_bytes;
+    if ((num_bytes = read(socket, buf + read_bytes,
+                          sizeof(buf) - read_bytes - 1)) == -1) {
+      return "";
+    }
+    read_bytes += num_bytes;
+  } while (read_bytes < content_bytes);
+
+  buf[content_bytes] = '\0';
+
+  return std::string{buf};
+}
+
+void CurlTestUtils::StartMockServer() {
+  // Get the socket file descriptor
+  int listening_socket = socket(AF_INET, SOCK_STREAM, 0);
+
+  // Create the socket address object
+  // The port is set to 0, meaning that it will be auto assigned
+  // Only local connections can access this socket
+  sockaddr_in socket_address{AF_INET, 0, htonl(INADDR_LOOPBACK)};
+  socklen_t socket_address_size = sizeof(socket_address);
+  if (listening_socket == -1) {
+    return;
+  }
+
+  // Bind the file descriptor to the socket address object
+  if (bind(listening_socket, (sockaddr*)&socket_address, socket_address_size) ==
+      -1) {
+    return;
+  }
+
+  // Assign an available port to the socket address object
+  if (getsockname(listening_socket, (sockaddr*)&socket_address,
+                  &socket_address_size) == -1) {
+    return;
+  }
+
+  // Get the port number
+  port_ = ntohs(socket_address.sin_port);
+
+  // Set server_thread_ operation to socket listening
+  server_thread_ = std::thread([=] {
```

Comment:
the lambda is quite big.
Declaring it as standalone function might be better (e.g void ServerLoop(int listening_socket)).